src/proficiency.h

```diff
@@ -40,6 +48,13 @@ class proficiency
         time_duration _time_to_learn = 9999_hours;
         std::set<proficiency_id> _required;
 
+        std::string _category;
+
+        int _str_bonus = 0;
+        int _dex_bonus = 0;
+        int _int_bonus = 0;
+        int _per_bonus = 0;
```

Comment:
Could this instead be some sort of container, e.g.
```
struct bonus_value {
    proficiency_bonus_list type;
    int value;
};

std::map<std::string, std::vector<bonus_value>> bonuses;
```
Then the JSON for this would look like `"bonuses": { "Archery": [ { "type": "strength", "value": 2 } ] }`

proficiency_bonus_list should probably get a different name, but this is much easier to expand on than the current setup.

The API doesn't need to change, but it's a lot more flexible.

Reply:
Would you happen to have an example of loading in the values from JSON for the above format? I'm having trouble getting this off the ground since I'm not familiar at all with this

Reply:
This should roughly do, let me know if you have further questions.
Comments are for illustrating how things work, they aren't necessarily things that should be documented here.
```diff
index 76a755f012..1e4c468086 100644
--- a/src/proficiency.cpp
--- a/src/proficiency.cpp
+++ b/src/proficiency.cpp
@@ -42,6 +42,42 @@ void proficiency::reset()
     proficiency_factory.reset();
 }

+template<>
+struct enum_traits<proficiency_bonus_list> {
+    static constexpr proficiency_bonus_list last = proficiency_bonus_list::last;
+};
+
+namespace io
+{
+
+template<>
+std::string enum_to_string<proficiency_bonus_list>( proficiency_bonus_list data )
+{
+    switch( data ) {
+        case proficiency_bonus_list::str_bonus:
+            return "str";
+        case proficiency_bonus_list::dex_bonus:
+            return "dex";
+        case proficiency_bonus_list::int_bonus:
+            return "per";
+        case proficiency_bonus_list::per_bonus:
+            return "int";
+        case proficiency_bonus_list::last:
+            break;
+    }
+    debugmsg( "Invalid proficiency bonus type" );
+    // Most other uses of this call abort() here. I don't think crashing the game is necessary.
+    return "";
+}
+} // namespace io
+
+void bonus_value::deserialize( JsonIn &jsin )
+{
+    const JsonObject &jo = jsin.get_object();
+
+    // The enum_to_string above handles this
+    mandatory( jo, false, "type", type );
+    mandatory( jo, false, "value", value );
+}
+
 void proficiency::load( const JsonObject &jo, const std::string & )
 {
     mandatory( jo, was_loaded, "name", _name );
@@ -53,12 +89,7 @@ void proficiency::load( const JsonObject &jo, const std::string & )
     optional( jo, was_loaded, "time_to_learn", _time_to_learn );
     optional( jo, was_loaded, "required_proficiencies", _required );

-    optional( jo, was_loaded, "category", _category );
-
-    optional( jo, was_loaded, "str_bonus", _str_bonus );
-    optional( jo, was_loaded, "dex_bonus", _dex_bonus );
-    optional( jo, was_loaded, "int_bonus", _int_bonus );
-    optional( jo, was_loaded, "per_bonus", _per_bonus );
+    // Our JSON handler can load maps with string keys as objects with the strings as the keys
+    // As long as the values are understood by it.
+    // Vectors innately are understood as JSON arrays
+    // And the deserialize function above tells it how to handle bonus_value s
+    optional( jo, was_loaded, "bonuses", _bonuses );
 }

 const std::vector<proficiency> &proficiency::get_all()
 ```
 ```diff
 diff --git a/src/proficiency.h b/src/proficiency.h
index 9cfcb39d64..0bddf3ccd7 100644
--- a/src/proficiency.h
+++ b/src/proficiency.h
@@ -23,11 +23,20 @@ struct learning_proficiency;
 template<typename T>
 class generic_factory;

-enum class proficiency_bonus_list : int {
+enum class proficiency_bonus_list : int {
     str_bonus,
     dex_bonus,
     int_bonus,
-    per_bonus
+    per_bonus,
+    last
+};
+
+
+struct bonus_value {
+    proficiency_bonus_list type;
+    int value;
+
+    void deserialize( JsonIn &jsin );
 };

 class proficiency
@@ -48,12 +57,7 @@ class proficiency
         time_duration _time_to_learn = 9999_hours;
         std::set<proficiency_id> _required;

-        std::string _category;
-
-        int _str_bonus = 0;
-        int _dex_bonus = 0;
-        int _int_bonus = 0;
-        int _per_bonus = 0;
+        std::map<std::string, std::vector<bonus_value>> _bonuses;

     public:
         static void load_proficiencies( const JsonObject &jo, const std::string &src );
```